cpp/src/glm/qn/qn.cuh

```diff
@@ -15,82 +15,70 @@
  */
 
 #pragma once
+
 #include "glm_base.cuh"
 #include "glm_linear.cuh"
 #include "glm_logistic.cuh"
 #include "glm_regularizer.cuh"
 #include "glm_softmax.cuh"
 #include "glm_svm.cuh"
 #include "qn_solvers.cuh"
+#include "qn_util.cuh"
+
+#include <cuml/linear_model/qn.h>
+
 #include <raft/matrix/math.hpp>
 #include <rmm/device_uvector.hpp>
 
 namespace ML {
 namespace GLM {
+
 template <typename T, typename LossFunction>
 int qn_fit(const raft::handle_t& handle,
+           const qn_params& pams,
            LossFunction& loss,
            const SimpleMat<T>& X,
            const SimpleVec<T>& y,
            SimpleDenseMat<T>& Z,
-           T l1,
-           T l2,
-           int max_iter,
-           T grad_tol,
-           T change_tol,
-           int linesearch_max_iter,
-           int lbfgs_memory,
-           int verbosity,
            T* w0_data,  // initial value and result
            T* fx,
            int* num_iters,
            cudaStream_t stream)
 {
-  LBFGSParam<T> opt_param;
-  opt_param.epsilon = grad_tol;
-  if (change_tol > 0) opt_param.past = 10;  // even number - to detect zig-zags
-  opt_param.delta          = change_tol;
-  opt_param.max_iterations = max_iter;
-  opt_param.m              = lbfgs_memory;
-  opt_param.max_linesearch = linesearch_max_iter;
+  LBFGSParam<T> opt_param(pams);
   SimpleVec<T> w0(w0_data, loss.n_param);
 
   // Scale the regularization strenght with the number of samples.
-  l1 /= X.m;
-  l2 /= X.m;
+  T l1 = pams.penalty_l1;
+  T l2 = pams.penalty_l2;
+  if (pams.penalty_normalized) {
```

Comment:
Would it be possible to add a test for a solver that needs `penalty_normalized=false`? E.g. a LinearSVM test that would converge to a different solution than sklearn, if the penalty normalization is not set correctly? (The penalty_normalized=true branch is currently tested by [this test](https://github.com/rapidsai/cuml/blob/962d6f03690d27705ac9c61ebb971a29239754d0/python/cuml/test/test_linear_model.py#L330), which would fail if the penalty parameter is not normalized correctly.)

Reply:
I can add it slightly later for ElasticNet model, when I integrate the QN solver there?.. This will be the case where `penalty_normalized=false` is required.

Reply:
Ok, that sounds fine to add there.